### PR TITLE
Update index.d.ts

### PR DESCRIPTION
**Pull request by SoraYama**
add containPixel in interface ECharts.

### CODE REVIEW

**Review by AM-young-fun — CHANGES_REQUESTED** [review]
请替换掉日文引号毕竟不太常用

**Review by AM-young-fun — CHANGES_REQUESTED** [review]
graphic定义成interface而不是对象。未在types里定义的echarts已有功能会变得不可用。其他人的现有代码用不了，兼容性有问题。

**Comment by AntiMoron** [comment]
暂时应该把那几个interface定义为对象。毕竟你功能覆盖不全，一招强类型定义所有人的代码都中招了，能编译过吗？

**Comment by AntiMoron** [comment]
何况最近echarts4有大动作，好多修改。

File: types/echarts/index.d.ts [hunk 1]
```diff
@@ -9,13 +9,11 @@ declare namespace echarts {
         renderer?: string
     }): ECharts;
 
-    const graphic: {
+    interface graphic {
```
Comment (HerringtonDarkholme) [inline types/echarts/index.d.ts]: Please use CamelCase for interface.

Like demonstrated here. http://definitelytyped.org/guides/best-practices.html

File: types/echarts/index.d.ts [hunk 2]
```diff
@@ -28,6 +26,15 @@ declare namespace echarts {
     function registerMap(mapName: string, geoJson: Object, specialAreas?: Object): void;
 
     function registerTheme(themeName: string, theme: Object): void;
+    
+    interface mapObj {
+        // 地图的 geoJson 数据
```
Comment (HerringtonDarkholme) [inline types/echarts/index.d.ts]: Please consider using JSDoc style comment for field comments. Also, English should be preferred.

File: types/echarts/index.d.ts (hunk 1 above)
Comment (HerringtonDarkholme) [reply types/echarts/index.d.ts]: graphic is a value, not a type. You still need to declare a value like `const graphic: Graphic`

**Comment by hyq77766177** [comment]
@AntiMoron Interface doesn't differ with object literal type in freshness check. Also, TypeScript uses a structural type system and also has an open interface. assign a value of type with additional properties to an interface type is totally acceptable, notwithstanding freshness check.
see: https://www.typescriptlang.org/play/#src=declare%20function%20knownProp(a%3A%20%7B%20test%3A%20number%20%7D)%3A%20void%0A%0AknownProp(%7B%0A%20%20%20%20test%3A%20123%2C%0A%20%20%20%20ssss%3A%20444%0A%7D)

**Comment by AntiMoron** [comment]
@SoraYama 你没get我的意思。所谓的structural type system就是一种duck typing罢了，typescript和js唯一的区别就是用这种“类型系统”限定了一层并加了一堆语法糖。interface限定了类型后，就会被typescript检查，别人调用你没在interface里定义的内容时就会报错（除非做cast），尽管他还是那个对象，但是被你的interface限定住有些线上代码是一定会受影响编译不过的。详情看例子：

https://www.typescriptlang.org/play/#src=const%20test%20%3D%20%7B%0A%20%20%20%20a%3A%201%2C%0A%20%20%20%20b%3A%202%0A%7D%3B%0A%0A%2F%2F%20%E5%81%87%E8%AE%BE%E4%B8%8B%E9%9D%A2%E8%BF%99%E8%A1%8C%E6%98%AF%E5%88%AB%E4%BA%BA%E5%B7%B2%E7%BB%8F%E5%86%99%E5%A5%BD%E4%B8%8A%E7%BA%BF%E4%BA%86%E7%9A%84%E4%BB%A3%E7%A0%81%0Aconsole.log(test.b)%3B%0A%0Ainterface%20Test%20%7B%20%0A%20%20%20%20a%3A%20number%0A%7D%0A%0A%2F%2F%20%E7%94%A8t%E4%BB%A3%E8%A1%A8%E8%A2%AB%E4%BD%A0%E7%94%A8interface%E9%99%90%E5%AE%9A%E8%BF%87%E4%BB%A3%E7%A0%81%0Aconst%20t%20%3D%20%3CTest%3Etest%3B%0A%2F%2F%20%E5%A5%BD%E4%BA%86%EF%BC%8C%E7%8E%B0%E5%9C%A8%E5%88%AB%E4%BA%BA%E7%9A%84%E7%BA%BF%E4%B8%8A%E4%BB%A3%E7%A0%81%E6%8A%A5%E9%94%99%E4%BA%86%0Aconsole.log(t.b)%3B

**Comment by SoraYama** [comment]
@AntiMoron 
If your type defined as `Object`, as your words, there's also an error will be triggered, see
https://www.typescriptlang.org/play/#src=%2F%2F%20%E5%A6%82%E6%9E%9C%E6%98%AF%E7%94%A8%E4%BD%A0%E8%AF%B4%E7%9A%84%20object%20%E8%80%8C%E4%B8%8D%E6%98%AF%20interface%0Aconst%20test%3A%20%7B%20a%3A%20number%3B%20%7D%20%3D%20%7B%0A%20%20%20%20a%3A%201%2C%0A%2F%2F%20%E4%B8%8D%E6%98%AF%E4%B8%80%E6%A0%B7%E7%BC%96%E8%AF%91%E4%B8%8D%E8%BF%87%0A%20%20%20%20b%3A%202%0A%7D%3B%0A
My work is just to pull out the duplicate definitions and rewrite, which won't break origin structure, I think. Btw, if `graphic` defined as a variable, it cannot be exported at all.

**Comment by AntiMoron** [comment]
@SoraYama  ```{ a: number; }``` 和 js里的```Object```是两个东西。你想要在ts里模拟js的Object应该用any. 所以你的例子应该写成

```typescript

// 如果是用你说的 object 而不是 interface
const test: any = {
    a: 1,
// 不是一样编译不过 <-- 怎么可能编译不过
    b: 2
};

```

**Comment by AntiMoron** [comment]
这个问题无外乎一个小朋友本来会画画会写字，你给他的个人介绍偏偏只介绍了会写字，结果班级集体活动安排任务老师无法安排画画的任务给他这样的道理。

**Comment by AntiMoron** [comment]
这么说把，我现在把你这个拿来我们线上工程甚至都编译不过。这代码再不修改我要回绝了。

**Comment by AntiMoron** [comment]
So... after judging with my colleague I decided to shutdown this merge request. Here are the reasons:

0. Compatibility. This code somehow may cause compile errors. Nobody can deny that define types explicitly is a good way to code. But as matter of fact Echarts is implemented in javascript. What we defined **here** is just a subset of origin functions. So compatibility comes to first. You shouldn't make other people waste their time to fix their worked code.
1. Stability. However you do make effort to contribute. But every modification for code review you add something new. This is really hard to judge its stability. 
2. Deletion. I saw some history code is deleted. I don't think this conforms to the OpenClosePrinciple.

Maybe later I will take care of your merge request. But for now let me reject & close it.

@HerringtonDarkholme  What's your opinion? Do you agree?

**Comment by HerringtonDarkholme** [comment]
> Compatibility. This code somehow may cause compile errors.

Can you provide an example that uses ECharts API and causes new error in updated declaration, but not in original one? I cannot see breaking changes here.

**Comment by AntiMoron** [comment]
@HerringtonDarkholme @SoraYama  My bad. I haven't seen the latest commit. It's ok. approved.Sorry for my misjudging.

**Review by AM-young-fun — APPROVED** [review]

**Comment by VirensCn** [comment]
@AntiMoron  按需加载怎么玩？？
'echarts/lib/echarts' 会报错，说没有这个。只能全部加载，但是全部加载后编译的文件起码大了300k

`import * as echarts from 'echarts/lib/echarts';

import 'echarts/lib/chart/bar';
import 'echarts/lib/chart/line';
import 'echarts/lib/component/legend';
import 'echarts/lib/component/toolbox';`

image

**Pull request merged**